README.md
# http-authentication
LDAP group - role mapping extension for the WordPress http-authentication plugin

## Description

This is an extension of the existing [http-authentication](https://wordpress.org/plugins/http-authentication) WordPress plugin which is not maintained any more.
The HTTP Authentication plugin allows you to use existing means of authenticating people to WordPress. Primarily it had been written and used for Apache’s basic HTTP authentication module.

![screenshot-1](screenshot-1.png?raw=true "Plugin options, allowing wordpress authentication")

Additionally you can configure an LDAP section where LDAP groups can be mapped to WordPress roles. The advantage against other plugins is that users are put into the right role **automatically** after login instead of editing every user separately. If no group matches for the user he/she will be put into the "New user default role" configured in the General section.

![screenshot-3](screenshot-3.png?raw=true "Plugin options for LDAP authentication")

## Installation

Clone this repository into your wordpress plugin directory (usually */usr/share/wordpress/wp-content/plugins* or */var/lib/wordpress/wp-content/plugins*), e.g.
```
git clone https://github.com/sunflowerbofh/http-authentication.git /usr/share/wordpress/wp-content/plugins
```

You can activate the plugin now within the Plugins menu.

## Configuration

After installation and activation you will find the item "HTTP Authentication" under the "Settings" button. 

### JSON

If you don't want to enter the settings manually, you can prepare a [json file](examples/http-authentication.json) and implement it with [wp cli](https://wp-cli.org/):
```
wp option update http_authentication_options --allow-root --path=/usr/share/wordpress --url=wordpress.example.com --format=json < example.json
```

### Web server

With the help of this plugin you can loop the user login through the webserver directly into wordpress. Here is an example how apache2 could be configured for LDAP:
```
<LocationMatch /(wp-login|wp-admin)>
    AuthType Basic
    AuthName "Enter Credentials"
    AuthBasicProvider ldap file
    AuthLDAPRemoteUserAttribute uid
    AuthLDAPRemoteUserIsDN off
    AuthLDAPGroupAttribute memberUid
    AuthLDAPGroupAttributeIsDN Off
    AuthLDAPURL "ldaps://ldap.example.com:636/dc=bfh?uid?sub"
    <RequireAny>
        Require ldap-group cn=wordpress-admin,ou=security,ou=groups,dc=example
        Require ldap-group cn=wordpress-editor,ou=security,ou=groups,dc=example
        Require ldap-group cn=wordpress-author,ou=security,ou=groups,dc=example
    </RequireAny>
</LocationMatch>
```
If you want a local admin user beside you can additionally include a file. Create a htpasswd with you wpadmin before.

```
    <RequireAny>
        Require ldap-group cn=wordpress,ou=security,ou=groups,dc=example
        AuthUserFile /etc/wordpress/htpasswd
        Require user wpadmin
    </RequireAny>
```

## ToDo List / Backlog

The LDAP frontend is still pretty bare-bone. The following enhancements are desirable:
* The plugin was only developped for "pure" LDAP, Active Directory should work as well but is not tested yet.
* Missing "Check LDAP connection" button for easier debugging the ldap connection.
* LDAP groups have to be entered fully manually. An autocompletion by getting results from the LDAP server would be nice to have.
* Instead of fixed fields lines should be added by '+'. These should contain a text box for the LDAP groups and a dropdown for the wordpress role.
* Pre-defined values are not visible (e.g. LDAP version = 3).
* Before filling out the fields and saving for the first time, you can get some ugly warnings about "non-defined index".


CHANGELOG.txt
2022-01-10	Katharina Drexel <katharina.drexel@bfh.ch>

	* Releasing version 4.6.

	[Katharina Drexel]
	* Adding original version (https://wordpress.org/plugins/http-authentication/).
	* Adding license, license header and README.
	* Adding changes for php>=7.0
	* Adding LDAP group role mapping.
	* Adding screenshot with LDAP examples.


options-page.php
<?php

/*
Copyright (C) 2011-2012 Daniel Westermann-Clark <daniel@danieltwc.com>
Copyright (C) 2022 Katharina Drexel <katharina.drexel@bfh.ch>

SPDX-License-Identifier: GPL-2.0+

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

class HTTPAuthenticationOptionsPage {
	var $plugin;
	var $group;
	var $page;
	var $options;
	var $title;

	function __construct($plugin, $group, $page, $options, $title = 'HTTP Authentication') {
		$this->plugin = $plugin;
		$this->group = $group;
		$this->page = $page;
		$this->options = $options;
		$this->title = $title;

		add_action('admin_init', array($this, 'register_options'));
		add_action('admin_menu', array($this, 'add_options_page'));
	}

	/*
	 * Register the options for this plugin so they can be displayed and updated below.
	 */
	function register_options() {
		register_setting($this->group, $this->group, array($this, 'sanitize_settings'));

		$section = 'http_authentication_main';
		add_settings_section($section, 'Main Options', array($this, '_display_options_section'), $this->page);
		add_settings_field('http_authentication_allow_wp_auth', 'Allow WordPress authentication?', array($this, '_display_option_allow_wp_auth'), $this->page, $section, array('label_for' => 'http_authentication_allow_wp_auth'));
		add_settings_field('http_authentication_auth_label', 'Authentication label', array($this, '_display_option_auth_label'), $this->page, $section, array('label_for' => 'http_authentication_auth_label'));
		add_settings_field('http_authentication_login_uri', 'Login URI', array($this, '_display_option_login_uri'), $this->page, $section, array('label_for' => 'http_authentication_login_uri'));
		add_settings_field('http_authentication_logout_uri', 'Logout URI', array($this, '_display_option_logout_uri'), $this->page, $section, array('label_for' => 'http_authentication_logout_uri'));
		add_settings_field('http_authentication_additional_server_keys', '$_SERVER variables', array($this, '_display_option_additional_server_keys'), $this->page, $section, array('label_for' => 'http_authentication_additional_server_keys'));
		add_settings_field('http_authentication_auto_create_user', 'Automatically create accounts?', array($this, '_display_option_auto_create_user'), $this->page, $section, array('label_for' => 'http_authentication_auto_create_user'));
		add_settings_field('http_authentication_auto_create_email_domain', 'Email address domain', array($this, '_display_option_auto_create_email_domain'), $this->page, $section, array('label_for' => 'http_authentication_auto_create_email_domain'));
		$section2 = 'ldap_authentication_main';
		add_settings_section($section2, 'LDAP Options', array($this, '_display_ldap_options_section'), $this->page);
		add_settings_field('http_authentication_allow_ldap', 'Activate LDAP group-role mapping?', array($this, '_display_option_allow_ldap'), $this->page, $section2, array('label_for' => 'http_authentication_allow_ldap'));
		add_settings_field('http_authentication_ldap_protocol', 'LDAP protocol (usually ldap or ldaps)', array($this, '_display_option_ldap_protocol'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_protocol'));
		add_settings_field('http_authentication_ldap_server', 'LDAP server (FQDN)', array($this, '_display_option_ldap_server'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_server'));
		add_settings_field('http_authentication_ldap_port', 'LDAP port', array($this, '_display_option_ldap_port'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_port'));
		add_settings_field('http_authentication_ldap_version', 'LDAP version', array($this, '_display_option_ldap_version'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_version'));
		add_settings_field('http_authentication_ldap_user', 'LDAP user', array($this, '_display_option_ldap_user'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_user'));
		add_settings_field('http_authentication_ldap_password', 'LDAP password', array($this, '_display_option_ldap_password'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_password'));
		add_settings_field('http_authentication_ldap_search_base', 'LDAP search base', array($this, '_display_option_ldap_search_base'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_search_base'));
		add_settings_field('http_authentication_ldap_group_dn', 'LDAP group DN', array($this, '_display_option_ldap_group_dn'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_group_dn'));
		add_settings_field('http_authentication_ldap_admin_group', 'LDAP administrator group', array($this, '_display_option_ldap_admin_group'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_admin_group'));
		add_settings_field('http_authentication_ldap_editor_group', 'LDAP editor group', array($this, '_display_option_ldap_editor_group'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_editor_group'));
		add_settings_field('http_authentication_ldap_author_group', 'LDAP author group', array($this, '_display_option_ldap_author_group'), $this->page, $section2, array('label_for' => 'http_authentication_ldap_author_group'));
	}

	/*
	 * Set the database version on saving the options.
	 */
	function sanitize_settings($input) {
		$output = $input;
		$output['db_version'] = $this->plugin->db_version;
		$output['allow_wp_auth'] = isset($input['allow_wp_auth']) ? (bool) $input['allow_wp_auth'] : false;
		$output['allow_ldap'] = isset($input['allow_ldap']) ? (bool) $input['allow_ldap'] : false;
		$output['auto_create_user'] = isset($input['auto_create_user']) ? (bool) $input['auto_create_user'] : false;

		return $output;
	}

	/*
	 * Add an options page for this plugin.
	 */
	function add_options_page() {
		add_options_page($this->title, $this->title, 'manage_options', $this->page, array($this, '_display_options_page'));
	}

	/*
	 * Display the options for this plugin.
	 */
	function _display_options_page() {
		if (! current_user_can('manage_options')) {
			wp_die(__('You do not have sufficient permissions to access this page.'));
		}
?>
<div class="wrap">
  <h2>HTTP Authentication Options</h2>
  <p>For the Login URI and Logout URI options, you can use the following variables to support your installation:</p>
  <ul>
    <li><code>%host%</code> - The current value of <code>$_SERVER['HTTP_HOST']</code></li>
    <li><code>%base%</code> - The base domain URL (everything before the path)</li>
    <li><code>%site%</code> - The WordPress home URI</li>
    <li><code>%redirect%</code> - The return URI provided by WordPress</li>
  </ul>
  <p>You can also use <code>%host_encoded%</code>, <code>%site_encoded%</code>, and <code>%redirect_encoded%</code> for URL-encoded values.</p>
  <form action="options.php" method="post">
    <?php settings_errors(); ?>
    <?php settings_fields($this->group); ?>
    <?php do_settings_sections($this->page); ?>
    <p class="submit">
      <input type="submit" name="Submit" value="<?php esc_attr_e('Save Changes'); ?>" class="button-primary" />
    </p>
  </form>
</div>
<?php
	}

	/*
	 * Display explanatory text for the main options section.
	 */
	function _display_options_section() {
	}
	/*
	 * Display explanatory text for the ldap options section.
	 */
	function _display_ldap_options_section() {
?>
Mapping roles to LDAP groups (optional). Works only if the wordpress server can connect to an LDAP server. If no group matches user is put into default role.
<?php
	}

	/*
	 * Display the WordPress authentication checkbox.
	 */
	function _display_option_allow_wp_auth() {
		$allow_wp_auth = $this->options['allow_wp_auth'];
		$this->_display_checkbox_field('allow_wp_auth', $allow_wp_auth);
?>
Should the plugin fallback to WordPress authentication if none is found from the server?
<?php
		if ($allow_wp_auth && $this->options['login_uri'] == htmlspecialchars_decode(wp_login_url())) {
			echo '<br /><strong>WARNING</strong>: You must set the login URI below to your external authentication system. Otherwise you will not be able to login!';
		}
	}

	/*
	 * Display the authentication label field, describing the authentication system
	 * in use.
	 */
	function _display_option_auth_label() {
		$auth_label = $this->options['auth_label'];
		$this->_display_input_text_field('auth_label', $auth_label);
?>
Default is <code>HTTP authentication</code>; override to use the name of your single sign-on system.
<?php
	}

	/*
	 * Display the login URI field.
	 */
	function _display_option_login_uri() {
		$login_uri = $this->options['login_uri'];
		$this->_display_input_text_field('login_uri', $login_uri);
?>
Default is <code><?php echo wp_login_url(); ?></code>; override to direct users to a single sign-on system. See above for available variables.<br />
Example: <code>%base%/Shibboleth.sso/Login?target=%redirect_encoded%</code>
<?php
	}

	/*
	 * Display the logout URI field.
	 */
	function _display_option_logout_uri() {
		$logout_uri = $this->options['logout_uri'];
		$this->_display_input_text_field('logout_uri', $logout_uri);
?>
Default is <code><?php echo htmlspecialchars(remove_query_arg('_wpnonce', htmlspecialchars_decode(wp_logout_url()))); ?></code>; override to e.g. remove a cookie. See above for available variables.<br />
Example: <code>%base%/Shibboleth.sso/Logout?return=%redirect_encoded%</code>
<?php
	}

	/*
	 * Display the additional $_SERVER keys field.
	 */
	function _display_option_additional_server_keys() {
		$additional_server_keys = $this->options['additional_server_keys'];
		$this->_display_input_text_field('additional_server_keys', $additional_server_keys);
?>
<code>$_SERVER</code> variables in addition to <code>REMOTE_USER</code> and <code>REDIRECT_REMOTE_USER</code> to check for the username value, separated by a comma. Use this to e.g. support personal X.509 certificates for authentication.<br />
Example: <code>SSL_CLIENT_S_DN_CN</code>
<?php
	}

	/*
	 * Display the automatically create accounts checkbox.
	 */
	function _display_option_auto_create_user() {
		$auto_create_user = $this->options['auto_create_user'];
		$this->_display_checkbox_field('auto_create_user', $auto_create_user);
?>
Should a new user be created automatically if not already in the WordPress database?<br />
Created users will obtain the role defined under &quot;New User Default Role&quot; on the <a href="options-general.php">General Options</a> page.
<?php
	}

	/*
	 * Display the email domain field.
	 */
	function _display_option_auto_create_email_domain() {
		$auto_create_email_domain = $this->options['auto_create_email_domain'];
		$this->_display_input_text_field('auto_create_email_domain', $auto_create_email_domain);
?>
When a new user logs in, this domain is used for the initial email address on their account. The user can change his or her email address by editing their profile.
<?php
	}

	/*
	 * Display the LDAP mapping checkbox.
	 */
	function _display_option_allow_ldap() {
		$allow_ldap = $this->options['allow_ldap'];
		$this->_display_checkbox_field('allow_ldap', $allow_ldap);
?>
Activate ldap group to role mapping?
<?php
		if ($allow_ldap ) {
			echo '<br /><strong>Hint:</strong> You must set at least the LDAP server and define one group.';
		}
	}

	/*
	 * Display the LDAP protocol field
	 */
	function _display_option_ldap_protocol() {
		$ldap_protocol = $this->options['ldap_protocol'];
		$this->_display_input_text_field('ldap_protocol', $ldap_protocol);
?>
Default is <code>ldaps</code>; use <code>ldap</code> when the server has no encryption.
<?php
	}

	/*
	 * Display the LDAP server field
	 */
	function _display_option_ldap_server() {
		$ldap_server = $this->options['ldap_server'];
		$this->_display_input_text_field('ldap_server', $ldap_server);
?>
Fully qualified LDAP server name (e.g. ldap.example.com)
<?php
	}

	/*
	 * Display the LDAP server field
	 */
	function _display_option_ldap_port() {
		$ldap_port = $this->options['ldap_port'];
		$this->_display_input_text_field('ldap_port', $ldap_port);
?>
LDAP port, usually 389 or 636
<?php
	}

	/*
	 * Display the LDAP version field
	 */
	function _display_option_ldap_version() {
		$ldap_version = $this->options['ldap_version'];
		$this->_display_input_text_field('ldap_version', $ldap_version);
?>
LDAP version, mostly you can leave it at 3.
<?php
	}

	/*
	 * Display the LDAP user field
	 */
	function _display_option_ldap_user() {
		$ldap_user = $this->options['ldap_user'];
		$this->_display_input_text_field('ldap_user', $ldap_user);
?>
User for LDAP bind, leave empty for anonymous bind.
<?php
	}

	/*
	 * Display the LDAP password field
	 */
	function _display_option_ldap_password() {
		$ldap_password = $this->options['ldap_password'];
		$this->_display_input_password_field('ldap_password', $ldap_password);
?>
Password for LDAP user, leave empty for anonymous bind.
<?php
	}

	/*
	 * Display the LDAP search base field
	 */
	function _display_option_ldap_search_base() {
		$ldap_search_base = $this->options['ldap_search_base'];
		$this->_display_input_text_field('ldap_search_base', $ldap_search_base);
?>
LDAP search base
<?php
	}

	/*
	 * Display the LDAP group dn field
	 */
	function _display_option_ldap_group_dn() {
		$ldap_group_dn = $this->options['ldap_group_dn'];
		$this->_display_input_text_field('ldap_group_dn', $ldap_group_dn);
?>
LDAP group DN. Will be concatenated with LDAP groups below (can be left empty when you want to use LDAP groups from different trees; in that case you have to enter the full path within each ldap group field).
<?php
	}

	/*
	 * Display the LDAP admin group field
	 */
	function _display_option_ldap_admin_group() {
		$ldap_admin_group = $this->options['ldap_admin_group'];
		$this->_display_input_text_field('ldap_admin_group', $ldap_admin_group);
?>
LDAP group which should be mapped to the role 'administrator'. Can be full LDAP tree or only cn (if LDAP group field is set). Leave empty when you don't want that role.
<?php
	}

	/*
	 * Display the LDAP editor group field
	 */
	function _display_option_ldap_editor_group() {
		$ldap_editor_group = $this->options['ldap_editor_group'];
		$this->_display_input_text_field('ldap_editor_group', $ldap_editor_group);
?>
LDAP group which should be mapped to the role 'editor' (leave empty when you don't want that role).
<?php
	}

	/*
	 * Display the LDAP author group field
	 */
	function _display_option_ldap_author_group() {
		$ldap_author_group = $this->options['ldap_author_group'];
		$this->_display_input_text_field('ldap_author_group', $ldap_author_group);
?>
LDAP group which should be mapped to the role 'author' (leave empty when you don't want that role).
<?php
	}

	/*
	 * Display a text input field.
	 */
	function _display_input_text_field($name, $value, $size = 75) {
?>
<input type="text" name="<?php echo htmlspecialchars($this->group); ?>[<?php echo htmlspecialchars($name); ?>]" id="http_authentication_<?php echo htmlspecialchars($name); ?>" value="<?php echo htmlspecialchars($value) ?>" size="<?php echo htmlspecialchars($size); ?>" /><br />
<?php
	}

	/*
	 * Display a password input field.
	 */
	function _display_input_password_field($name, $value, $size = 75) {
?>
<input type="password" name="<?php echo htmlspecialchars($this->group); ?>[<?php echo htmlspecialchars($name); ?>]" id="http_authentication_<?php echo htmlspecialchars($name); ?>" value="<?php echo htmlspecialchars($value) ?>" size="<?php echo htmlspecialchars($size); ?>" /><br />
<?php
	}

	/*
	 * Display a checkbox field.
	 */
	function _display_checkbox_field($name, $value) {
?>
<input type="checkbox" name="<?php echo htmlspecialchars($this->group); ?>[<?php echo htmlspecialchars($name); ?>]" id="http_authentication_<?php echo htmlspecialchars($name); ?>"<?php if ($value) echo ' checked="checked"' ?> value="1" /><br />
<?php
	}
}
?>


http-authentication.php
<?php
/*
Plugin Name: HTTP Authentication
Version: 4.6
Plugin URI: https://github.com/sunflowerbofh/http-authentication
Description: Authenticate users using basic HTTP authentication (<code>REMOTE_USER</code>). This plugin assumes users are externally authenticated, as with <a href="http://www.gatorlink.ufl.edu/">GatorLink</a>. Additionally, if using LDAP, put users of certain LDAP groups in corresponding wordpress roles.
Author: Katharina Drexel
Author URI: https://github.com/sunflowerbofh
*/

/* Copyright (C) 2011-2012 Daniel Westermann-Clark <daniel@danieltwc.com>
Copyright (C) 2022 Katharina Drexel <katharina.drexel@bfh.ch>

SPDX-License-Identifier: GPL-2.0+

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 2 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.
*/

require_once(dirname(__FILE__) . DIRECTORY_SEPARATOR . 'options-page.php');

class HTTPAuthenticationPlugin {
	var $db_version = 2;
	var $option_name = 'http_authentication_options';
	var $options;

	function __construct() {
		$this->options = get_option($this->option_name);

		if (is_admin()) {
			$options_page = new HTTPAuthenticationOptionsPage($this, $this->option_name, __FILE__, $this->options);
			add_action('admin_init', array($this, 'check_options'));
		}

		add_action('login_head', array($this, 'add_login_css'));
		add_action('login_footer', array($this, 'add_login_link'));
		add_action('check_passwords', array($this, 'generate_password'), 10, 3);
		add_action('wp_logout', array($this, 'logout'));
		add_filter('login_url', array($this, 'bypass_reauth'));
		add_filter('show_password_fields', array($this, 'allow_wp_auth'));
		add_filter('allow_password_reset', array($this, 'allow_wp_auth'));
		add_filter('authenticate', array($this, 'authenticate'), 10, 3);
	}

	/*
	 * Check the options currently in the database and upgrade if necessary.
	 */
	function check_options() {
		if ($this->options === false || ! isset($this->options['db_version']) || $this->options['db_version'] < $this->db_version) {
			if (! is_array($this->options)) {
				$this->options = array();
			}

			$current_db_version = isset($this->options['db_version']) ? $this->options['db_version'] : 0;
			$this->upgrade($current_db_version);
			$this->options['db_version'] = $this->db_version;
			update_option($this->option_name, $this->options);
		}
	}

	/*
	 * Upgrade options as needed depending on the current database version.
	 */
	function upgrade($current_db_version) {
		$default_options = array(
			'allow_wp_auth' => false,
			'auth_label' => 'HTTP authentication',
			'login_uri' => htmlspecialchars_decode(wp_login_url()),
			'logout_uri' => remove_query_arg('_wpnonce', htmlspecialchars_decode(wp_logout_url())),
			'additional_server_keys' => '',
			'auto_create_user' => false,
			'auto_create_email_domain' => '',
			'allow_ldap' => false,
			'ldap_protocol' => 'ldaps',
			'ldap_server' => 'ldap.example.com',
			'ldap_port' => '636',
			'ldap_version' => 3,
			'ldap_user' => '',
			'ldap_password' => '',
			'ldap_search_base' => 'dc=example',
			'ldap_group_dn' => 'ou=groups,dc=example',
			'ldap_admin_group' => '',
			'ldap_editor_group' => '',
			'ldap_author_group' => '',
		);

		if ($current_db_version < 1) {
			foreach ($default_options as $key => $value) {
				// Handle migrating existing options from before we stored a db_version
				if (! isset($this->options[$key])) {
					$this->options[$key] = $value;
				}
			}
		}
	}

	function add_login_css() {
?>
<style type="text/css">
p#http-authentication-link {
  width: 100%;
  height: 4em;
  text-align: center;
  margin-top: 2em;
}
p#http-authentication-link a {
  margin: 0 auto;
  float: none;
}
</style>
<?php
	}

	/*
	 * Add a link to the login form to initiate external authentication.
	 */
	function add_login_link() {
		global $redirect_to;

		$login_uri = $this->_generate_uri($this->options['login_uri'], wp_login_url($redirect_to));
		$auth_label = $this->options['auth_label'];

		echo "\t" . '<p id="http-authentication-link"><a class="button-primary" href="' . htmlspecialchars($login_uri) . '">Log In with ' . htmlspecialchars($auth_label) . '</a></p>' . "\n";
	}

	/*
	 * Generate a password for the user. This plugin does not require the
	 * administrator to enter this value, but we need to set it so that user
	 * creation and editing works.
	 */
	function generate_password($username, $password1, $password2) {
		if (! $this->allow_wp_auth()) {
			$password1 = $password2 = wp_generate_password();
		}
	}

	/*
	 * Logout the user by redirecting them to the logout URI.
	 */
	function logout() {
		$logout_uri = $this->_generate_uri($this->options['logout_uri'], home_url());

		wp_redirect($logout_uri);
		exit();
	}

	/*
	 * Remove the reauth=1 parameter from the login URL, if applicable. This allows
	 * us to transparently bypass the mucking about with cookies that happens in
	 * wp-login.php immediately after wp_signon when a user e.g. navigates directly
	 * to wp-admin.
	 */
	function bypass_reauth($login_url) {
		$login_url = remove_query_arg('reauth', $login_url);

		return $login_url;
	}

	/*
	 * Can we fallback to built-in WordPress authentication?
	 */
	function allow_wp_auth() {
		return (bool) $this->options['allow_wp_auth'];
	}

	/*
	 * Authenticate the user, first using the external authentication source.
	 * If allowed, fall back to WordPress password authentication.
	 */
	function authenticate($user, $username, $password) {
		$user = $this->check_remote_user();

		if (! is_wp_error($user)) {
			// User was authenticated via REMOTE_USER
			$user = new WP_User($user->ID);
		}
		else {
			// REMOTE_USER is invalid; now what?

			if (! $this->allow_wp_auth()) {
				// Bail with the WP_Error when not falling back to WordPress authentication
				wp_die($user);
			}

			// Fallback to built-in hooks (see wp-includes/user.php)
		}

		return $user;
	}

	/*
	 * If the REMOTE_USER or REDIRECT_REMOTE_USER evironment variable is set, use it
	 * as the username. This assumes that you have externally authenticated the user.
	 */
	function check_remote_user() {
		$username = '';

		$server_keys = $this->_get_server_keys();
		foreach ($server_keys as $server_key) {
			if (! empty($_SERVER[$server_key])) {
				$username = $_SERVER[$server_key];
			}
		}

		if (! $username) {
			return new WP_Error('empty_username', '<strong>ERROR</strong>: No user found in server variables.');
		}

		// Create new users automatically, if configured
		$user = get_user_by('login', $username);
		if (! $user)  {
			if ((bool) $this->options['auto_create_user']) {
				$user = $this->_create_user($username);
			}
			else {
				// Bail out to avoid showing the login form
				$user = new WP_Error('authentication_failed', __('<strong>ERROR</strong>: Invalid username or incorrect password.'));
			}
		}

		return $user;
	}

	/*
	 * Use ldap mapping?
	 */
	function allow_ldap() {
		return (bool) $this->options['allow_ldap'];
	}

	/*
	* Set the corresponding role to user
	 */
	function group_role_mapping($user) {
		$ldapuser = $user;
		echo "<br /> User is " . $ldapuser ."\n";

		$ldap_protocol = $this->options['ldap_protocol'];
		$ldap_server = $this->options['ldap_server'];
		$ldap_port = $this->options['ldap_port'];
		$ldap_version = $this->options['ldap_version'];
		$ldap_user = $this->options['ldap_user'];
		$ldap_password = $this->options['ldap_password'];
		$ldap_search_base = $this->options['ldap_search_base'];
		$ldap_group_dn = $this->options['ldap_group_dn'];
		$ldap_admin_group = $this->options['ldap_admin_group'];
		$ldap_editor_group = $this->options['ldap_editor_group'];
		$ldap_author_group = $this->options['ldap_author_group'];

		$ldap_uri = $ldap_protocol . "://" . $ldap_server . ':' . $ldap_port;
		#if (empty($ldap_version)) {
		#	$ldap_version = 3;
		#}
		$ldapconn = ldap_connect($ldap_uri) or wp_die(__('Could not connect to LDAP server.'));
		ldap_set_option($ldapconn, LDAP_OPT_PROTOCOL_VERSION, $ldap_version);
		if( $ldapconn ) {
			if ( empty($ldap_user)) {
				$ldapbind = ldap_bind( $ldapconn ) or wp_die (__("ERROR trying to bind: " .ldap_error($ldapconn)));
			} else {
				$ldapbind = ldap_bind( $ldapconn, $ldap_user, $ldap_password ) or wp_die (__("ERROR trying to bind: " .ldap_error($ldapconn)));
			}

			if ( $ldapbind ) {
				$group_roles = array( $ldap_admin_group=>"administrator", $ldap_editor_group=>"editor", $ldap_author_group=>"author" );
				foreach ( $group_roles as $key=>$value ) {
					if ( ! empty( $key )) {
						if ( empty( $ldap_group_dn )) {
							$filter = "(&(uid=". $ldapuser . ")(memberOf=cn=" . $key . "))";
						} else {
							$filter = "(&(uid=". $ldapuser . ")(memberOf=cn=" . $key . "," . $ldap_group_dn . "))";
						}
						$result = ldap_search ($ldapconn, $ldap_search_base, $filter) or wp_die (__("ERROR in search query: " .ldap_error($ldapconn)));
						$data = ldap_get_entries($ldapconn, $result);

						if ( ! empty($data[0]) ) {
							if ( $data[0]['uid'][0] === $ldapuser ) {
								$user_role = $value;
								break;
							}
						}
					}
				}
			} else {
				wp_die(__("LDAP bind failed"));
			}

		}
	return $user_role;
		# if (! $this->allow_ldap()) -> $role = call function($user)
		# if ! empty($role)
		# $user->set_role( $role ); (nach $user = new \WP_User( $user_id );)
	}

	/*
	 * Return the list of $_SERVER keys that we will check for a username. By
	 * default, these are REMOTE_USER and REDIRECT_REMOTE_USER. Additional keys
	 * can be configured from the options page.
	 */
	function _get_server_keys() {
		$server_keys = array('REMOTE_USER', 'REDIRECT_REMOTE_USER');

		$additional_server_keys = $this->options['additional_server_keys'];
		if (! empty($additional_server_keys)) {
			$keys = preg_split('/,\s*/', $additional_server_keys);
			$server_keys = array_merge($server_keys, $keys);
		}

		return $server_keys;
	}

	/*
	 * Create a new WordPress account for the specified username.
	 */
	function _create_user($username) {
		$password = wp_generate_password();
		$email_domain = $this->options['auto_create_email_domain'];

		$user_id = wp_create_user($username, $password, $username . ($email_domain ? '@' . $email_domain : ''));
		$user = get_user_by('id', $user_id);

		return $user;
	}

	/*
	 * Fill the specified URI with the site URI and the specified return location.
	 */
	function _generate_uri($uri, $redirect_to) {
		// Support tags for staged deployments
		$base = $this->_get_base_url();

		$tags = array(
			'host' => $_SERVER['HTTP_HOST'],
			'base' => $base,
			'site' => home_url(),
			'redirect' => $redirect_to,
		);

		foreach ($tags as $tag => $value) {
			$uri = str_replace('%' . $tag . '%', $value, $uri);
			$uri = str_replace('%' . $tag . '_encoded%', urlencode($value), $uri);
		}

		// Support previous versions with only the %s tag
		if (strstr($uri, '%s') !== false) {
			$uri = sprintf($uri, urlencode($redirect_to));
		}

		return $uri;
	}

	/*
	 * Return the base domain URL based on the WordPress home URL.
	 */
	function _get_base_url() {
		$home = parse_url(home_url());

		$base = home_url();
		foreach (array('path', 'query', 'fragment') as $key) {
			if (! isset($home[$key])) continue;
			$base = str_replace($home[$key], '', $base);
		}

		return $base;
	}
}

// Load the plugin hooks, etc.
$http_authentication_plugin = new HTTPAuthenticationPlugin();
?>
